algorithms/src/snark/marlin/data_structures/proof.rs

```diff
@@ -131,99 +134,136 @@ impl<F: PrimeField> Evaluations<F> {
+                    z_b_evals_collect.insert(circuit_id, values);
+                }
+            } else if label.contains("g_a") {
+                g_a_evals.push(*value);
+            } else if label.contains("g_b") {
+                g_b_evals.push(*value);
+            } else if label.contains("g_c") {
+                g_c_evals.push(*value);
+            }
+        }
+        let z_b_evals = z_b_evals_collect.into_values().collect();
+        Self { z_b_evals, g_1_eval: map["g_1"], g_a_evals, g_b_evals, g_c_evals }
     }
 
-    pub(crate) fn get(&self, label: &str) -> Option<F> {
-        if let Some(index) = label.strip_prefix("z_b_") {
-            self.z_b_evals.get(index.parse::<usize>().unwrap()).copied()
+    pub(crate) fn get(&self, circuit_index: usize, label: &str) -> Option<F> {
+        if label == "g_1" {
+            return Some(self.g_1_eval);
+        }
+
+        if let Some(index) = label.find("z_b_") {
+            let z_b_eval_circuit = &self.z_b_evals[circuit_index];
+            let instance_index = label[index + 4..].parse::<usize>().unwrap();
+            z_b_eval_circuit.get(instance_index).copied()
+        } else if label.contains("g_a") {
+            self.g_a_evals.get(circuit_index).copied()
+        } else if label.contains("g_b") {
+            self.g_b_evals.get(circuit_index).copied()
+        } else if label.contains("g_c") {
+            self.g_c_evals.get(circuit_index).copied()
         } else {
-            match label {
-                "g_1" => Some(self.g_1_eval),
-                "g_a" => Some(self.g_a_eval),
-                "g_b" => Some(self.g_b_eval),
-                "g_c" => Some(self.g_c_eval),
-                _ => None,
-            }
+            None
         }
     }
+
+    pub fn to_field_elements(&self) -> Vec<F> {
+        let mut result: Vec<F> = self.z_b_evals.clone().into_iter().flatten().collect();
+        result.extend([self.g_1_eval]);
+        result.extend(self.g_a_evals.clone());
+        result.extend(self.g_b_evals.clone());
+        result.extend(self.g_c_evals.clone());
+        result
+    }
 }
 
 impl<F: PrimeField> Valid for Evaluations<F> {
     fn check(&self) -> Result<(), snarkvm_utilities::SerializationError> {
         self.z_b_evals.check()?;
         self.g_1_eval.check()?;
-        self.g_a_eval.check()?;
-        self.g_b_eval.check()?;
-        self.g_c_eval.check()
-    }
-}
-
-impl<F: PrimeField> Evaluations<F> {
-    pub fn to_field_elements(&self) -> Vec<F> {
-        let mut result = self.z_b_evals.clone();
-        result.extend([self.g_1_eval, self.g_a_eval, self.g_b_eval, self.g_c_eval]);
-        result
+        self.g_a_evals.check()?;
+        self.g_b_evals.check()?;
+        self.g_c_evals.check()
     }
 }
 
 /// A zkSNARK proof.
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct Proof<E: PairingEngine> {
     /// The number of instances being proven in this proof.
-    batch_size: usize,
+    batch_sizes: Vec<usize>,
```

Comment:
Perhaps we should explicitly serialize this as a `u64`? (I know we changed `usize` to serialize to a `u64` in #1413, but best to be sure, I think.

Reply:
Yes great idea was pondering it as well